atomicfu/src/concurrentTest/kotlin/kotlinx/atomicfu/locks/ThreadParkerTest.kt

```diff
@@ -0,0 +1,57 @@
+package kotlinx.atomicfu.locks
+
+import kotlinx.atomicfu.atomic
+import kotlin.test.Test
+import kotlin.test.assertTrue
+import kotlin.time.Duration
+
+class ThreadParkerTest {
+
+
+    @Test
+    fun parkUnpark() {
+        var parkingHandle: ParkingHandle? = null
+        var done = false
+        
+        testThread { 
+            parkingHandle = ParkingSupport.currentThreadHandle()
+            ParkingSupport.park(Duration.INFINITE)
+            done = true
+        }
+        
+        // Allow thread to be parked before park call
+        sleepMillis(100)
+        ParkingSupport.unpark(parkingHandle!!)
+        
+        // Allow thread to wakeup
+        sleepMillis(400)
+        assertTrue(done)
+    }
+    
+
+    @Test
+    fun unparkPark() {
+        var parkingHandle: ParkingHandle? = null
+        var done = false
+
+        testThread {
+            parkingHandle = ParkingSupport.currentThreadHandle()
+            
+            // Allow main thread to pre-unpark
+            sleepMillis(200)
```

Comment:
This delay can also be avoided by introducing `@Volatile var unparked = false` that's set to `true` after `unpark`, and here, do `while (!unparked) { sleepMillis(10) }`.